grug/malachite/src/actors/host/host.rs

```diff
@@ -0,0 +1,560 @@
+                    height,
+                    round,
+                    validator_address,
+                    value_bytes,
+                    reply_to,
+                )?;
+
+                Ok(())
+            },
+            HostMsg::ExtendVote { reply_to, .. } => {
+                reply_to.send(None)?;
+                Ok(())
+            },
+            HostMsg::VerifyVoteExtension { reply_to, .. } => {
+                reply_to.send(Ok(()))?;
+                Ok(())
+            },
+        }
+    }
+
+    fn consensus_ready(
+        &self,
+        state: &mut State,
+        consensus: ConsensusRef<Context>,
+    ) -> ActorResult<()> {
+        let height = DECIDED_BLOCK
+            .keys(state, None, None, grug::Order::Descending)
+            .next()
+            .transpose()?
+            .unwrap_or_default();
+
+        state.set_consensus(consensus.clone());
+
+        // TODO: is a sleep necessary here? It's present in the example implementation
+
+        consensus.cast(ConsensusMsg::StartHeight(
+            <ctx!(Height)>::new(height + 1),
+            self.validator_set.clone(),
+        ))?;
+
+        Ok(())
+    }
+
+    fn get_history_min_height(
+        &self,
+        state: &State,
+        reply_to: RpcReplyPort<ctx!(Height)>,
+    ) -> ActorResult<()> {
+        let height = DECIDED_BLOCK
+            .keys(state, None, None, grug::Order::Ascending)
+            .next()
+            .transpose()?
+            .unwrap_or_default();
+
+        reply_to.send(<ctx!(Height)>::new(height))?;
+
+        Ok(())
+    }
+
+    fn started_round(
+        &self,
+        state: &mut State,
+        height: ctx!(Height),
+        round: Round,
+        proposer: ctx!(Address),
+        role: Role,
+    ) -> ActorResult<()> {
+        state.started_round(height, round, proposer, role);
+
+        // If we have already built or seen one or more values for this height and round,
+        // feed them back to consensus. This may happen when we are restarting after a crash.
+        let consensus = state.consensus()?;
+
+        for value in UNDECIDED_PROPOSALS
+            .prefix(*height)
+            .append(round.as_i64())
+            .values(state, None, None, grug::Order::Ascending)
+        {
+            let value = value?;
+            info!(%height, %round, hash = ?value.value, "Replaying already known proposed value");
+
+            consensus.cast(ConsensusMsg::ReceivedProposedValue(
+                value,
+                ValueOrigin::Consensus,
+            ))?;
+        }
+        Ok(())
+    }
+
+    async fn restream_value(
+        &self,
+        state: &mut State,
+        _height: ctx!(Height),
+        _round: Round,
+        valid_round: Round,
+        _address: ctx!(Address),
+        block_hash: ctx!(Value::Id),
+    ) -> ActorResult<()> {
+        if let Some(data) = UNDECIDED_BLOCK.may_load(state, block_hash)? {
+            // TODO: Should we assert height, round and proposer?
```

Comment:
For height is fine to assert.
Why do you need a round in the block? If this is a re-proposal of a valid value you need to assert on `valid_round`. When restream is triggered by the liveness property (`valid_round` is not set), you would assert on `round`. 
Proposer may be different than the one you have in a block so don't assert on that.